src/amber/router/pipe/csrf.cr

```diff
@@ -16,20 +15,23 @@ module Amber
         @session_key = "csrf.token"
         @header_key = "HTTP_X_CSRF_TOKEN"
         @param_key = "_csrf"
-        @check_methods = CHECK_METHODS
       end
 
       def call(context : HTTP::Server::Context)
-        if !check_methods.includes?(context.request.method) || valid_token?(context)
+        if !CHECK_METHODS.includes?(context.request.method) || valid_token?(context)
           call_next(context)
         else
           raise Amber::Exceptions::Forbidden.new("CSRF check failed.")
         end
       end
 
       def valid_token?(context)
-        context.params.fetch(param_key, nil) == token(context) ||
-          context.request.headers.fetch(header_key, nil) == token(context)
+        if context.params[param_key]? == token(context) || context.request.headers[header_key]? == token(context)
```

Comment:
If we ever had to write a unit test for this method it would be difficult since the method is doing 2 things. Also I would expect the method to change anything base on the name for those reasons I think you should move`context.session.delete(session_key)` to line 22. 

```crystal
context.session.delete(session_key)
call_next(context)
```